### DISCUSSION TITLE
Write File in Azure Blob

### DISCUSSION THREAD

**Discussion by akash-goel** [vaexio/vaex#1997] [Q&A]
Hi Team ,

We are using Azure blob Storage and want to write a file in Blob using Vaex . we have tried export_hdf5 and export Functions it did not worked for us , Can you please suggest how we can write a hdf5 file in Azure Blob usign Vaex.

Thanks

**Comment by akash-goel** [comment]
Do Vaex supports Azure Blob Storage or any way to write file in blob storage using Vaex ? Please guide

**Comment by JovanVeljanoski** [comment]
Take a look at:
- https://github.com/vaexio/vaex/issues/1272
- https://github.com/vaexio/vaex/issues/1248

Should probably help you out

**Reply by akash-goel** [reply]
Thanks Jovan for checking this request .

we are able to read the file from Blob , we are facing issue with write operation only.
When we are using export_hdf5 function and passing filesystem then getting this error "TypeError: export_hdf5(url,fs) got an unexpected keyword argument 'fs'"

If we are not passing filesystem then export_hdf5 function not working and getting this error 
`FileNotFoundError: [Errno 2] Unable to create file (unable to open file: name = 'abfs://jupytorhub/test.hdf5', errno = 2, error message = 'No such file or directory', flags = 13, o_flags = 242) 
`

As per the documentation export_hdf5 is not supporting fileystem , Request you to guide is there any way to write a file in Azure Blob

**Reply by JovanVeljanoski** [reply]
Try `export_arrow` or `export_parquet` where other file systems are supported.

**Reply by akash-goel** [reply]
Thanks Jovan

we have tried with export_arrow ,export_parquet and export_many option as well , for 20 million records Vaex is taking 80 seconds to read the data but during write operation into the Azure blob for both (Parquet and Arrow ) it is generating the files but taking more than 30 min. If we are using Dask Library then it is taking 5 min only . Is this expected time from Vaex or anything we are missing , please find the below code your reference.
Request you to guide us in case we are missing anything .

Parquet Generation Code

%%time
```python
import vaex
import adlfs
from azure.storage.blob import BlobServiceClient, generate_blob_sas, BlobSasPermissions, __version__
engagement_code = 'E-T123456789'
product_type = 'Type'
account_name = 'accountName'        
account_key = 'accountKey'                
container_name = 'container'
fs = adlfs.AzureBlobFileSystem(account_name='accountName', account_key='accountKey')
mergedDf.export_many('containerName/output_changed_chunk_arrow/output_chunk-{i:03}.parquet', chunk_size=1000000,fs=fs)
```

Export Arrow 
```python
%%time
import vaex
import adlfs
from azure.storage.blob import BlobServiceClient, generate_blob_sas, BlobSasPermissions, __version__
engagement_code = 'E-T123456789'
product_type = 'Type'
account_name = 'accountName'        
account_key = 'accountKey'                
container_name = 'container'
fs = adlfs.AzureBlobFileSystem(account_name='accountName', account_key='accountKey')
mergedDf.export_many('containerName/output_changed_chunk_arrow/output_chunk-{i:03}.parquet', chunk_size=1000000,fs=fs)
```

**Reply by JovanVeljanoski** [reply]
Can you format your code nice please.. hard to read like this

**Reply by akash-goel** [reply]
Done

**Reply by maartenbreddels** [reply]
> When we are using export_hdf5 function and passing filesystem then getting this error "TypeError: export_hdf5(url,fs) got an unexpected keyword argument 'fs'"

hdf5 cannot be written to the cloud, only read (see https://vaex.io/docs/example_io.html#Cloud-Support )

How are you using dask here, is this multi node or multi processing?
